relm4/src/shared_state/reducer.rs

```diff
@@ -0,0 +1,247 @@
+use std::sync::{Arc, RwLock};
+
+use once_cell::sync::Lazy;
+
+use crate::{Sender, RUNTIME};
+
+use super::SubscriberFn;
+
+/// A trait that implements a reducer function.
+///
+/// For more information, see [`Reducer`].
+pub trait Reducible {
+    /// The input message type used to modify the data.
+    type Input;
+
+    /// Initialize the data.
+    fn init() -> Self;
+
+    /// Process the input message and update the state.
+    ///
+    /// Return [`true`] to notify all subscribers.
+    /// Return [`false`] to ignore all subscribers.
+    ///
+    /// For example, it makes sense to return [`false`] to indicate
+    /// that the message had no (noteworthy) effect on the data and
+    /// the subscribers don't need to be notified.
+    fn reduce(&mut self, input: Self::Input) -> bool;
+}
+
+struct ReducerInner<Data: Reducible> {
+    sender: Sender<Data::Input>,
+    subscribers: Arc<RwLock<Vec<SubscriberFn<Data>>>>,
+}
+
+impl<Data> Default for ReducerInner<Data>
+where
+    Data: Reducible + Send + 'static,
+    Data::Input: Send,
+{
+    fn default() -> Self {
+        let (sender, receiver) = crate::channel();
+        let subscribers: Arc<RwLock<Vec<SubscriberFn<Data>>>> = Arc::default();
+
+        let rt_subscribers = subscribers.clone();
+        RUNTIME.spawn(async move {
+            let mut data = Data::init();
+            while let Some(input) = receiver.recv().await {
+                if data.reduce(input) {
+                    rt_subscribers
+                        .read()
+                        .unwrap()
+                        .iter()
+                        .for_each(|subscriber| subscriber(&data));
+                }
+            }
+        });
+
+        Self {
+            sender,
+            subscribers,
+        }
+    }
+}
+
+impl<Data> std::fmt::Debug for ReducerInner<Data>
+where
+    Data: std::fmt::Debug + Reducible,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ReducerInner")
+            .field("sender", &self.sender)
+            .field("subscribers", &self.subscribers.try_read().map(|s| s.len()))
+            .finish()
+    }
+}
+
+/// A type that allows you to share information across your
+/// application easily.
+///
+/// Reducers receiver messages, update their state accordingly
+/// and the notify their subscribers.
+///
+/// Unlike [`SharedState`](super::SharedState), this type works with
+/// messages and an update function.
+/// It's not possible to access the internal data directly.
+/// You can only modify by sending messages and read in subscriber functions.
```

Comment:
I'm having trouble understanding this sentence.

Reply:
Should be a lot better now :)